### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by kcwoolverton**
Merging current, stable dev branch into master

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -1,21 +0,0 @@
-MIT License
```
Comment (Omicron3141) [inline LICENSE]: Does this file need to be removed? If it was an accident, we should probably reset it on dev.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,24 @@
-# Classroom Dashboard - KBOP
-This is the KBOP team version of Classroom Dashboard, a classroom engagement tool written for Harvey Mudd College's CS121
+# README
```
Comment (Omicron3141) [inline README.md]: We might want to fill this out, or at least revert the change to what we had before.

File: app/assets/stylesheets/application.css [hunk 3]
```diff
@@ -1,15 +0,0 @@
-/*
```
Comment (Omicron3141) [inline app/assets/stylesheets/application.css:1]: Out of curiosity, why did we switch from css to scss?

File: app/assets/stylesheets/basic.css [hunk 4]
```diff
@@ -0,0 +1,39 @@
+/*
```
Comment (Omicron3141) [inline app/assets/stylesheets/basic.css:1]: If I recall, this is the CSS for the bulk upload zone? If so, could we add a comment explaining why it needs to be there and more about where it came from?

File: app/assets/stylesheets/dropzone.css [hunk 5]
```diff
@@ -0,0 +1,388 @@
+/*
```
Comment (Omicron3141) [inline app/assets/stylesheets/dropzone.css:1]: Would it be possible to add a comment detailing how this file differs from / interacts with basic.css above?

File: app/controllers/courses_controller.rb [hunk 6]
```diff
@@ -0,0 +1,103 @@
+class CoursesController < ApplicationController    
+    def index
+        @user = User.find(params[:user_id])
+        proper_user(@user)
```
Comment (Omicron3141) [inline app/controllers/courses_controller.rb]: Is it worth changing the name of this function to check_proper_user() or ensure_authorized_user() or something like that? I feel like it might make the code a bit more self-explanatory.

File: app/controllers/students_controller.rb [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        if @student.update(student_params)
+            redirect_to user_course_path(@user, @course)
+        else
+            render 'edit'
+        end
+    end
+
+    # Handles batch upload for students.
+    # The name of the file (minus the file extension) will be the name of the student.
+    def import
```
Comment (Omicron3141) [inline app/controllers/students_controller.rb]: Is this function called once per student, or once per upload? Can we make that a bit more explicit in either the function name or the comments?

File: app/views/courses/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,32 @@
+<div class="container-fluid">
+  <div class="row">
+      <div class="col-lg-8 col-md-6 col-sm-4"><h1>Listing Courses</h1></div>
+      <div class="col-lg-4 col-md-6 col-sm-8" style="text-align:right;"><h1><%= link_to "New Course", new_user_course_path(@user), :class => "btn"%></h1></div>
```
Comment (Omicron3141) [inline app/views/courses/index.html.erb:4]: Out of curiosity, why is the text-alight:right a style entry instead of inside an alight-right class or something like that? It just seems a bit inconsistent, epecially when this might be something we do frequently.

File: app/views/courses/index.html.erb [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  </div>
+
+<table class="table table-hover table-responsive">
+  <thead >
+    <tr class= "row"; style="background-color:rgb(78,55,100,.5);">
```
Comment (Omicron3141) [inline app/views/courses/index.html.erb:9]: Same as above, with the background color.

File: app/views/courses/show.html.erb [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+      <h1> <%= @course.code %>: <%= @course.title %></h1>
+      <ul class="nav nav-pills nav-justified">
+        <li class = "active"><%= link_to "Course", user_course_path(@user, @course) %><br></li>
+        <li><%= link_to "Edit Course", edit_user_course_path(@user, @course) %><br></li>
+        <li><%= link_to "Student Roster", user_course_students_path(@user, @course) %><br></li>
+        <li><%= link_to "Face/Name Quiz", {controller:"courses", action:"quiz"} %><br></li>
+      </ul>
+      <div class = "container">
+        <div class="row">
+          <div class = "col-lg-4"></div>
```
Comment (Omicron3141) [inline app/views/courses/show.html.erb:22]: Is this div empty on purpose? If not, is it worth replacing it with a col-offset on the next div?

File: config/application.rb [hunk 11]
```diff
@@ -1,6 +1,7 @@
 require_relative 'boot'
 
 require 'rails/all'
+require 'csv'
```
Comment (Omicron3141) [inline config/application.rb]: Is this nessecary now that we're doing bulk upload by pictures?

**Review by Omicron3141 — CHANGES_REQUESTED** [review]
Overall, I think we're in good shape. The license and readme both got messed up and should probably be fixed, but most of the rest of my comments are kind of nitpicky.

File: app/assets/stylesheets/application.scss [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+    background-attachment: fixed;
+  }
+
+.row {
+  align-content: center;
+}
+
+.col-lg-10 {
+  align-content: center;
+}
```
Comment (mcshen) [inline app/assets/stylesheets/application.scss]: Nit: there should probably be a new line at the end of this file

File: app/views/courses/edit.html.erb [hunk 13]
```diff
@@ -0,0 +1,8 @@
+<center>
+    <br>
+    <div>
+        <h1>Edit Course</h1>
+        <hr>
+        <%= render partial: 'form', locals: {cancel_path: user_course_path(@user, @course), cancel_text: "Cancel"} %>
+    </div>
+</center>
```
Comment (mcshen) [inline app/views/courses/edit.html.erb]: I don't actually know why or when GitHub wants newlines at the end of files, but it says there should be one here, and in a few other files

File: app/views/layouts/application.html.erb [hunk 14]
```diff
@@ -1,11 +1,73 @@
...
+        <%= link_to "Log Out", "/logout", method: :delete %>
+        </li>
+      </ul>
+    </div>
+  </div>
+</nav>
+
+
+<% else %>
+<nav class="navbar navbar-default">
```
Comment (mcshen) [inline app/views/layouts/application.html.erb:47]: The indentation seems off here. I think everything within this else statement should be indented more to match the if statement.

File: Gemfile [hunk 15]
```diff
@@ -13,11 +13,18 @@ gem 'sqlite3'
 # Use Puma as the app server
 gem 'puma', '~> 3.7'
 # Use SCSS for stylesheets
-gem 'sass-rails', '~> 5.0'
+gem 'sass-rails', '>= 3.2'
```
Comment (wynnee) [inline Gemfile:16]: So, the change from 5.0 down to 3.2 was an effort to debug and get sass/bootstrap working. In case we run into errors with 3.2 or want to use 5.0, this will be the place to change it.

File: test/controllers/courses_controller_test.rb [hunk 16]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class CoursesControllerTest < ActionDispatch::IntegrationTest
+  # test "the truth" do
+  #   assert true
+  # end
```
Comment (mcshen) [inline test/controllers/courses_controller_test.rb:6]: What are these comments for? Were we supposed to add tests here?

File: Gemfile [hunk 17]
```diff
@@ -28,7 +35,15 @@ gem 'jbuilder', '~> 2.5'
 # Use Redis adapter to run Action Cable in production
 # gem 'redis', '~> 3.0'
 # Use ActiveModel has_secure_password
-# gem 'bcrypt', '~> 3.1.7'
+gem 'bcrypt', '~> 3.1.11'
+
+gem 'bootstrap-sass', '~> 3.3.7'
+
+gem 'jquery-rails'
```
Comment (wynnee) [inline Gemfile:42]: Jquery was necessary for Bootstrap - I am not currently using it for anything else.

File: README.md (hunk 2 above)
Comment (wynnee) [reply README.md]: We definitely will need to flesh out our README, it's fine now but it should be more thorough.

File: app/assets/javascripts/application.js [hunk 18]
```diff
@@ -11,5 +11,7 @@
 // about supported directives.
 //
 //= require rails-ujs
+//= require jquery
+//= require bootstrap-sprockets
```
Comment (wynnee) [inline app/assets/javascripts/application.js:15]: For Bootstrap

File: app/views/homepage/index.html.erb [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+        <% if flash.any? %>
+          <% flash.each do |type, message| %>
+            <div class="alert alert-<%= type %>"><%= message %></div>
+          <% end %>
+        <% end %>
+        <%= yield %>
+      </div>
+    </div>
+  </body>
+</html
```
Comment (mcshen) [inline app/views/homepage/index.html.erb]: In some documents, like this one, end tags are aligned with the open tags, but in other documents the end tags are less indented by one space. We should probably be more consistent.

File: app/assets/javascripts/users.coffee [hunk 20]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (wynnee) [inline app/assets/javascripts/users.coffee]: do we need to keep these coffee files? Is anyone using them>?

File: app/assets/stylesheets/application.css (hunk 3 above)
Comment (wynnee) [reply app/assets/stylesheets/application.css:1]: Sass uses scss. We could use css, but scss is more powerful/I really wanted to try it. Here's a link to a good explanation of why I chose it over flat css: [https://alistapart.com/article/why-sass](here)

File: app/assets/stylesheets/application.scss (hunk 12 above)
Comment (wynnee) [inline app/assets/stylesheets/application.scss]: I think this is pretty self-explanatory, but I can explain this file if anyone wants me to! This is definitely not the finished stylesheet. I need to clean up the row and col-lg-10 parts though- that was part of trying to fix some rows.

File: app/assets/stylesheets/basic.css (hunk 4 above)
Comment (wynnee) [reply app/assets/stylesheets/basic.css:1]: I have not yet worked on this css yet, comments would be appreciated so I can understand what I'll be working with.

File: app/assets/stylesheets/dropzone.css (hunk 5 above)
Comment (wynnee) [reply app/assets/stylesheets/dropzone.css:1]: Is this a file I should be touching for frontend or no?

File: app/assets/stylesheets/homepage.scss [hunk 21]
```diff
@@ -0,0 +1,10 @@
+// Place all the styles related to the Homepage controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
+@import "bootstrap-sprockets";
+@import "bootstrap";
+
+body {
+    background-color: rosybrowns;
+    font-family: proxima-nova, 'Helvetica Neue', Arial, Helvetica, sans-serif;
+  }
```
Comment (wynnee) [inline app/assets/stylesheets/homepage.scss]: I am unsure where this body scss comes from. I may delete it and see if it changes anything.

File: app/views/courses/index.html.erb (hunk 8 above)
Comment (wynnee) [reply app/views/courses/index.html.erb:4]: I believe @mcshen worked on this bit

File: app/views/courses/show.html.erb (hunk 10 above)
Comment (wynnee) [reply app/views/courses/show.html.erb:22]: I think that might have been left over from something one of use deleted but didn't get rid of the tags. Something to clean up.

**Review by wynnee — CHANGES_REQUESTED** [review]
I think there is some front end for me to clean up before we pull. Other than that, everything looks good. I have yet to make frontend for quiz/batch uploading, but I think that is next phase for me.

File: Gemfile (hunk 17 above)
Comment (mcshen) [reply Gemfile:42]: I'm pretty sure Jonah used it for something.

File: app/assets/stylesheets/dropzone.css (hunk 5 above)
Comment (Omicron3141) [reply app/assets/stylesheets/dropzone.css:1]: I think this is from Kate's ticket, so she might know more about where it came from

File: Gemfile (hunk 17 above)
Comment (Omicron3141) [reply Gemfile:42]: Yeah, I use JQuery in the flashcards and remote submissions now

File: app/assets/javascripts/users.coffee (hunk 20 above)
Comment (Omicron3141) [reply app/assets/javascripts/users.coffee]: I think it should be safe to get rid of them. They all should be auto-generated, and we'd need to replace them with .js files to do anything with them anyways

File: README.md (hunk 2 above)
Comment (Omicron3141) [reply README.md]: This new version looks good to me

**Review by Omicron3141 — CHANGES_REQUESTED** [review]
The new license and readme look good, but I think we should hold for the new comments.

File: config/application.rb (hunk 11 above)
Comment (kcwoolverton) [reply config/application.rb]: No, it is not. I will remove it.

File: app/controllers/courses_controller.rb (hunk 6 above)
Comment (kcwoolverton) [reply app/controllers/courses_controller.rb]: Good idea, I will make this change.

File: app/views/courses/edit.html.erb (hunk 13 above)
Comment (kcwoolverton) [reply app/views/courses/edit.html.erb]: Github wants newlines at the end of all files because the UNIX standard says to end each file with a newline--it is a generally accepted style thing, and in very, very rare situations involving multiple, consecutively run shell scripts, it can actually change functionality.

**Review by Omicron3141 — APPROVED** [review]

**Comment by Omicron3141** [comment]
I think we're basically good. May <diety> have mercy on our <life-orbs>

**Review by wynnee — APPROVED** [review]
